bt_ddos_shield/miner_shield.py

```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
         """
-        pass
+        if self.worker_thread is not None:
+            raise Exception("Shield is already enabled")
+
+        self.finishing = False
+        self.run = True
+        self._add_task(MinerShieldInitializeTask())
+        self.worker_thread = threading.Thread(target=self._worker_function)
+        self.worker_thread.start()
+
+    def disable(self):
+        """
+        Disable shield. It stops worker thread after finishing current task. Function blocks until worker is stopped.
+        """
+        self._add_task(MinerShieldDisableTask())
+        self.finishing = True
+        self.worker_thread.join()
+        self.worker_thread = None
+        self.task_queue = Queue() # clear task queue
 
     def ban_validator(self, validator_hotkey: Hotkey):
         """
-        Ban a validator by its hotkey. Function blocks execution until manifest file is updated and info about file
-        is published to Bittensor.
+        Ban a validator by its hotkey. Task will be executed by worker. It will update manifest file and publish info
+        about new file version to blockchain.
 
         Args:
             validator_hotkey: The hotkey of the validator.
         """
+        self._add_task(MinerShieldBanValidatorTask(validator_hotkey))
+        pass
+
+    def _add_task(self, task):
+        """
+        Add task to task queue. It will be handled by _worker_function.
+        """
+        if not isinstance(task, MinerShieldTask):
+            raise Exception("Task is not instance of MinerShieldTask")
+        if not self.run:
+            raise Exception("Shield is disabled")
+
+        self.task_queue.put(task)
+
+    def _worker_function(self):
+        """
+        Function called in separate thread by enable() to start the shield. It is handling events put to task_queue.
+        """
+
+        self.event_processor.add_event(MinerShieldEvent(f"Starting shield"))
+
+        while self.run:
+            task = self.task_queue.get()
+            try_count = 1
+
+            while self.run:
+                self.event_processor.add_event(MinerShieldEvent(f"Handling task {task}, try {try_count}"))
+
+                try:
+                    task.handle(self)
+                    self.event_processor.add_event(MinerShieldEvent(f"Task {task} finished successfully"))
+                    break
+                except Exception as e:
+                    self.event_processor.add_event(MinerShieldEvent(f"Error during handling task {task}", e))
+
+                    if self.finishing:
+                        break
+
+                    try_count += 1
+                    sleep(self.options.retry_delay)
+
+        self.event_processor.add_event(MinerShieldEvent(f"Stopping shield"))
+
+    def _handle_initialize(self):
+        """
+        Initialize shield. Load state and initial validators set.
+        """
+        self.state_manager.get_state()
+        self.event_processor.add_event(MinerShieldEvent("State loaded"))
```

Comment:
Rather than seeing
```py
self.event_processor.add_event(MinerShieldEvent("State loaded"))
```
everywhere, I'd prefer to see
```py
self._event("State loaded")
```

with
```py
def _event(self, description: str, *args, **kwargs):
    return self.event_processor.add_event(self.event_class(description, *args, **kwargs))
```
and `event_class=MinerShieldEvent` in the constructor